File: backend/src/main/kotlin/Parser.kt
import com.github.javaparser.JavaParser
import com.github.javaparser.ParserConfiguration
import com.github.javaparser.ParserConfiguration.LanguageLevel
import com.github.javaparser.Problem
import com.github.javaparser.ast.CompilationUnit
import java.nio.file.Path

class SourceFilesParser(languageLevel: LanguageLevel) {
    private val parser: JavaParser

    init {
        val config = ParserConfiguration()
        config.languageLevel = languageLevel
        parser = JavaParser(config)
    }

    fun parse(filenames: List<Path>, errorCallback: (List<Pair<Path, Problem>>) -> Nothing): Map<Path, CompilationUnit> {
        val parsingResults = filenames.map { it to parser.parse(it) }
        val problems = parsingResults.flatMap { (path, result) ->
            result.problems.map { problem -> path to problem }
        }
        return if (problems.isEmpty()){
            parsingResults.associate { it.first to it.second.result.get() }
        } else {
            errorCallback(problems)
        }
    }

}

File: s2sCompiler/src/main/scala/Parser.scala
package s2sCompiler

import com.github.javaparser.ParserConfiguration.LanguageLevel
import com.github.javaparser.ast.CompilationUnit
import com.github.javaparser.{JavaParser, ParserConfiguration}
import ErrorReporter.ErrorLevel.*

import com.github.javaparser.symbolsolver.JavaSymbolSolver
import com.github.javaparser.symbolsolver.resolution.typesolvers.{CombinedTypeSolver, JavaParserTypeSolver, ReflectionTypeSolver}

import java.io.File

final class Parser(srcDirectory: File) extends CompilerStage[File, CompilationUnit] {

  private val symbolSolver = new JavaSymbolSolver(new CombinedTypeSolver(new ReflectionTypeSolver(), new JavaParserTypeSolver(srcDirectory)))
  private val javaparser = new JavaParser(
    new ParserConfiguration()
      .setLanguageLevel(LanguageLevel.JAVA_17)
      .setSymbolResolver(symbolSolver)
  )

  override protected def runImpl(srcFile: File, errorReporter: ErrorReporter): Option[CompilationUnit] = {
    val parseRes = javaparser.parse(srcFile)
    if (parseRes.isSuccessful) {
      Some(parseRes.getResult.get())
    } else {
      parseRes.getProblems.forEach { problem =>
        errorReporter.reportErrorPos(problem.toString, NonFatalError, srcFile.getName, problem.getLocation.flatMap(_.toRange))
      }
      None
    }
  }

}
